File: core/Lang/MathIr.fs
namespace Server.Lang

// ---------------------------------------------------------------------------
// Math IR — F# mirror of pointer_mk18/kernel/src/math_ir.zig.
//
// Expression DAG for distance fields. Nodes form a topologically sorted array
// (each node references earlier ones by index). Affines, intrinsics, and
// sketch primitives live in side tables; nodes of kind RemapAffine /
// Intrinsic carry an index into those tables.
//
// Future MathIrCodec.fs serializes this to the wire format scene_decode.zig
// expects; keep enum ordering in lock-step with math_ir.zig so that mapping
// is a direct int cast.
// ---------------------------------------------------------------------------

module MathIr =

    type Axis =
        | X = 0
        | Y = 1
        | Z = 2

    type Plane =
        | XY = 0
        | XZ = 1
        | YZ = 2

    /// Order MUST match math_ir.zig's Unary enum (intFromEnum mapping).
    type Unary =
        | Neg = 0
        | Abs = 1
        | Recip = 2
        | Square = 3
        | Sqrt = 4
        | Floor = 5
        | Ceil = 6
        | Round = 7
        | Sin = 8
        | Cos = 9
        | Tan = 10
        | Asin = 11
        | Acos = 12
        | Atan = 13
        | Exp = 14
        | Ln = 15
        | Not = 16

    /// Order MUST match math_ir.zig's Binary enum.
    type Binary =
        | Add = 0
        | Sub = 1
        | Mul = 2
        | Div = 3
        | Atan2 = 4
        | Min = 5
        | Max = 6
        | Pow = 7
        | Compare = 8
        | Mod = 9
        | And_ = 10
        | Or_ = 11

    type NodeKind =
        | Var = 0
        | Slot = 1
        | Const = 2
        | UnaryK = 3
        | BinaryK = 4
        | RemapAxes = 5
        | RemapAffine = 6
        | Intrinsic = 7
        | Fold = 8
        // ── Sketch primitives as first-class subtree nodes ───────────────
        // Each one carries its coords / radius as child node refs via
        // `NodeRefs[A..A+B]`. The plane is encoded in `Op` (0=xy, 1=xz,
        // 2=yz). `ArcCenter` packs clockwise as the low bit of `Op`:
        // `Op = plane * 2 + (clockwise ? 1 : 0)`.
        | LineSegment      = 9
        | Circle           = 10
        | BezierQuadratic  = 11
        | BezierCubic      = 12
        | ArcCenter        = 13

    /// Variadic-fold operator. Stored as the `Op` field of a `Fold` node;
    /// the children are a `NodeRefs[A..A+B]` window of node ids.
    type FoldOp =
        | Min = 0
        | Max = 1
        | Sum = 2

    /// Intrinsics are kernel-level specialised evaluators. Packaging-only
    /// "min over a curve list" ops (formerly `SketchPath` / `SketchDistance`)
    /// were lowered to AST/MathIR `Fold` + per-primitive subtree nodes in
    /// Phases 2-3; `CurveDistanceAlong` stays as an intrinsic because its
    /// signed-distance-along-an-axis math is genuinely specialised.
    type IntrinsicKind =
        | CurveDistanceAlong = 2

    /// Reference to a node in MathIR.Nodes. -1 means "unset".
    type Expr = { Id: int }

    let nullExpr : Expr = { Id = -1 }

    type Vec2 = { x: float; y: float }
    type Vec3 = { x: float; y: float; z: float }
    type Interval = { Lo: float; Hi: float }
    type Box3 = { Xi: Interval; Yi: Interval; Zi: Interval }
    type Cube = { Center: Vec3; HalfSize: float }

    /// Mirrors math_ir.zig Node. `op` is the int value of the relevant enum
    /// (Axis for Var, slot id for Slot, Unary/Binary for ops, etc.).
    type Node = {
        Kind:  NodeKind
        Op:    int
        A:     int
        B:     int
        C:     int
        D:     int
        Value: float
    }

    let private freshNode kind = {
        Kind = kind; Op = 0; A = -1; B = -1; C = -1; D = -1; Value = 0.0
    }

    /// 3×4 affine: 3 rows × 4 columns (translation in the last column). Each
    /// matrix entry is a node reference, so slot- or constant-valued matrices
    /// fall out for free.
    type Affine3 = {
        M00: Expr; M01: Expr; M02: Expr; M03: Expr
        M10: Expr; M11: Expr; M12: Expr; M13: Expr
        M20: Expr; M21: Expr; M22: Expr; M23: Expr
    }

    // Old slot-backed `SketchPrimitive` record + `SlotPoint2` and the
    // `Primitives` side table were retired in Phase 3. Sketch primitives
    // are now first-class subtree node kinds (LineSegment / Circle /
    // BezierQuadratic / BezierCubic / ArcCenter), with their geometry
    // children packed into `NodeRefs`.

    type Intrinsic = {
        Kind: IntrinsicKind
        Plane: Plane
        PrimitiveStart: int
        PrimitiveCount: int
        Closed: bool
        Flip: bool
        Ox: int; Oy: int; Oz: int
        Ax: int; Ay: int; Az: int
    }

    /// Capacity caps mirror math_ir.zig. Not enforced at build time (we use
    /// ResizeArray); the codec will assert before serializing.
    let MaxNodes = 4096
    let MaxAffines = 256
    let MaxIntrinsics = 512
    let MaxTapeWords = 4096
    let MaxImmediates = 512

    /// Mutable IR builder. One instance per evaluation; the evaluator hands
    /// it to every builtin so they can append nodes / affines / intrinsics.
    type MathIR() =
        let nodes      = ResizeArray<Node>()
        let affines    = ResizeArray<Affine3>()
        let intrinsics = ResizeArray<Intrinsic>()
        /// Packed child-id array. Fold + sketch-primitive node kinds +
        /// CurveDistanceAlong all index into this via `(start, count)`
        /// windows. Replaces the legacy Primitives side table.
        let nodeRefs   = ResizeArray<int>()

        member _.Nodes      = nodes
        member _.Affines    = affines
        member _.Intrinsics = intrinsics
        member _.NodeRefs   = nodeRefs

        member _.Push(n: Node) : Expr =
            let id = nodes.Count
            nodes.Add(n)
            { Id = id }

        member this.Var(axis: Axis) : Expr =
            this.Push { freshNode NodeKind.Var with Op = int axis }

        member this.X() = this.Var Axis.X
        member this.Y() = this.Var Axis.Y
        member this.Z() = this.Var Axis.Z

        member this.SlotE(slotId: int) : Expr =
            this.Push { freshNode NodeKind.Slot with Op = slotId }

        member this.Constant(v: float) : Expr =
            this.Push { freshNode NodeKind.Const with Value = v }

        member this.Unary(op: Unary, a: Expr) : Expr =
            this.Push {
                freshNode NodeKind.UnaryK with Op = int op; A = a.Id
            }

        member this.Binary(op: Binary, a: Expr, b: Expr) : Expr =
            this.Push {
                freshNode NodeKind.BinaryK with Op = int op; A = a.Id; B = b.Id
            }

        member this.RemapAxes(target: Expr, x: Expr, y: Expr, z: Expr) : Expr =
            this.Push {
                freshNode NodeKind.RemapAxes with
                    A = target.Id; B = x.Id; C = y.Id; D = z.Id
            }

        member _.AddAffine(a: Affine3) : int =
            let id = affines.Count
            affines.Add(a)
            id

        member this.RemapAffine(target: Expr, affineId: int) : Expr =
            this.Push {
                freshNode NodeKind.RemapAffine with A = target.Id; B = affineId
            }

        // Sketch primitives now live as `*N` builders below
        // (LineSegmentN / CircleN / etc.) — they emit subtree nodes whose
        // child coords flow through the regular DAG.

        // ---- intrinsics --------------------------------------------------------

        member private _.PushIntrinsic(i: Intrinsic) : int =
            let id = intrinsics.Count
            intrinsics.Add(i)
            id

        /// Signed distance along `axis` to the closest of the given
        /// primitive curves. The primitives are stored as a contiguous
        /// window in `NodeRefs` — the same packing Fold uses — so a
        /// `CurveDistanceAlong` intrinsic carries no special-shape data
        /// beyond what the DAG can reach.
        ///
        /// `Intrinsic.PrimitiveStart` and `PrimitiveCount` are reused
        /// as the `NodeRefs` window. The names are vestigial — semantics
        /// are now "child-node window" — but kept to avoid wire-format
        /// churn.
        member this.CurveDistanceAlong(
                plane: Plane,
                primitives: Expr list,
                axisX: Expr,
                axisY: Expr,
                axisZ: Expr,
                flip: bool) : Expr =
            let start = nodeRefs.Count
            for p in primitives do
                nodeRefs.Add p.Id
            let count = nodeRefs.Count - start
            let i = {
                Kind = IntrinsicKind.CurveDistanceAlong
                Plane = plane
                PrimitiveStart = start
                PrimitiveCount = count
                Closed = false; Flip = flip
                Ox = -1; Oy = -1; Oz = -1
                Ax = axisX.Id; Ay = axisY.Id; Az = axisZ.Id
            }
            let id = this.PushIntrinsic i
            this.Push { freshNode NodeKind.Intrinsic with A = id }

        // ---- fold ---------------------------------------------------------

        /// Variadic fold over `children`. Stores the children's node ids in
        /// the packed `NodeRefs` array and pushes a `Fold` node whose `A`
        /// field names the start index and `B` the count. The runtime
        /// evaluator (F# eval, kernel tape) reads `NodeRefs[A..A+B]` and
        /// folds the children's values with `op`.
        ///
        /// Caller's responsibility to provide a non-empty list — a Fold over
        /// zero children has no well-defined identity here.
        member this.Fold(op: FoldOp, children: Expr list) : Expr =
            let start = nodeRefs.Count
            for child in children do
                nodeRefs.Add child.Id
            let count = nodeRefs.Count - start
            this.Push { freshNode NodeKind.Fold with Op = int op; A = start; B = count }

        // ---- sketch primitives as subtree nodes --------------------------
        //
        // Each primitive's coordinate inputs are stored as child node refs
        // in `NodeRefs[A..A+B]`. The plane is packed into `Op`. The
        // evaluators read each coord through normal node evaluation, so
        // endpoint coords can be any expression (Const, Slot, computed).

        member private this.PushPrimitiveNode
                (kind: NodeKind) (plane: Plane) (op_extra: int) (children: Expr list) : Expr =
            let start = nodeRefs.Count
            for child in children do
                nodeRefs.Add child.Id
            let count = nodeRefs.Count - start
            this.Push {
                freshNode kind with
                    Op = int plane * 2 + op_extra
                    A = start
                    B = count
            }

        /// 2D line segment from `(p0x, p0y)` to `(p1x, p1y)` evaluated on
        /// `plane`. Children order: p0x, p0y, p1x, p1y.
        member this.LineSegmentN(plane: Plane, p0x, p0y, p1x, p1y) : Expr =
            this.PushPrimitiveNode NodeKind.LineSegment plane 0 [ p0x; p0y; p1x; p1y ]

        /// Circle with center `(cx, cy)` and radius `r` on `plane`.
        /// Children order: cx, cy, r.
        member this.CircleN(plane: Plane, cx, cy, r) : Expr =
            this.PushPrimitiveNode NodeKind.Circle plane 0 [ cx; cy; r ]

        /// Quadratic Bézier. Children order: p0x, p0y, p1x, p1y, p2x, p2y.
        member this.BezierQuadraticN(plane: Plane, p0x, p0y, p1x, p1y, p2x, p2y) : Expr =
            this.PushPrimitiveNode NodeKind.BezierQuadratic plane 0
                [ p0x; p0y; p1x; p1y; p2x; p2y ]

        /// Cubic Bézier. Children order: p0x..p3y (8 children).
        member this.BezierCubicN(plane: Plane, p0x, p0y, p1x, p1y, p2x, p2y, p3x, p3y) : Expr =
            this.PushPrimitiveNode NodeKind.BezierCubic plane 0
                [ p0x; p0y; p1x; p1y; p2x; p2y; p3x; p3y ]

        /// Center-parameterised arc from `(sx, sy)` to `(ex, ey)` about
        /// `(cx, cy)`. `clockwise` is packed into the low bit of `Op`.
        /// Children order: sx, sy, ex, ey, cx, cy.
        member this.ArcCenterN(plane: Plane, sx, sy, ex, ey, cx, cy, clockwise: bool) : Expr =
            let extra = if clockwise then 1 else 0
            this.PushPrimitiveNode NodeKind.ArcCenter plane extra
                [ sx; sy; ex; ey; cx; cy ]

File: kernel/src/math_ir.zig
pub const Axis = enum(i32) { x, y, z };
pub const Plane = enum(i32) { xy, xz, yz };

pub const Unary = enum(i32) {
    neg,
    abs,
    recip,
    square,
    sqrt,
    floor,
    ceil,
    round,
    sin,
    cos,
    tan,
    asin,
    acos,
    atan,
    exp,
    ln,
    not,
};

pub const Binary = enum(i32) {
    add,
    sub,
    mul,
    div,
    atan2,
    min,
    max,
    pow,
    compare,
    mod,
    and_,
    or_,
};

pub const NodeKind = enum(i32) {
    var_,
    slot,
    const_,
    unary,
    binary,
    remap_axes,
    remap_affine,
    intrinsic,
    fold,
    // Sketch primitives as first-class subtree nodes. Each one's geometry
    // inputs are stored as child node refs in `NodeRefs[a..a+b]`. The
    // plane is packed in `op`: `op / 2` = plane (0=xy, 1=xz, 2=yz).
    // `arc_center` packs `clockwise` as the low bit of `op`.
    line_segment,
    circle,
    bezier_quadratic,
    bezier_cubic,
    arc_center,
};

/// Variadic-fold operator. Order must match `MathIr.FoldOp` on the F# side.
pub const FoldOp = enum(i32) {
    min,
    max,
    sum,
};

/// Kernel-level specialised evaluators. Packaging-only intrinsics
/// (formerly `sketch_distance` / `sketch_path`) were lowered to fold +
/// per-primitive subtree nodes; only `curve_distance_along` stays
/// intrinsic because its signed-distance-along-an-axis math is
/// genuinely specialised. Numeric values preserved for wire-format
/// compatibility: `sketch_distance=0`, `sketch_path=1`, `curve_distance_along=2`.
pub const IntrinsicKind = enum(i32) {
    curve_distance_along = 2,
};

pub const Expr = struct { id: i32 };
pub const Vec2 = struct { x: f64, y: f64 };
pub const Vec3 = struct { x: f64, y: f64, z: f64 };
pub const Interval = struct { lo: f64, hi: f64 };
pub const Box3 = struct { xi: Interval, yi: Interval, zi: Interval };
pub const Cube = struct { center: Vec3, half_size: f64 };
pub const Node = struct {
    kind: NodeKind,
    op: i32 = 0,
    a: i32 = -1,
    b: i32 = -1,
    c: i32 = -1,
    d: i32 = -1,
    value: f64 = 0.0,
};

pub const Affine3 = struct {
    m00: Expr,
    m01: Expr,
    m02: Expr,
    m03: Expr,
    m10: Expr,
    m11: Expr,
    m12: Expr,
    m13: Expr,
    m20: Expr,
    m21: Expr,
    m22: Expr,
    m23: Expr,
};

/// `CurveDistanceAlong`'s primitive children live as a window in
/// `node_refs[primitive_start..primitive_start+primitive_count]`. The
/// names retain "primitive" for wire-format and historical continuity,
/// but they always reference primitive subtree node ids now — not rows
/// in the (retired) `Primitives` side table.
pub const Intrinsic = struct {
    kind: IntrinsicKind,
    plane: Plane = .xy,
    primitive_start: i32 = -1,
    primitive_count: i32 = 0,
    closed: bool = false,
    flip: bool = false,
    ox: i32 = -1,
    oy: i32 = -1,
    oz: i32 = -1,
    ax: i32 = -1,
    ay: i32 = -1,
    az: i32 = -1,
};

pub const max_nodes = 4096;
pub const max_affines = 256;
pub const max_intrinsics = 512;
pub const max_node_refs = 8192;
pub const max_tape_words = 4096;
pub const max_immediates = 512;

pub const MathIR = struct {
    nodes: [max_nodes]Node = undefined,
    node_count: usize = 0,
    affines: [max_affines]Affine3 = undefined,
    affine_count: usize = 0,
    intrinsics: [max_intrinsics]Intrinsic = undefined,
    intrinsic_count: usize = 0,
    /// Packed child-id array. Used by Fold (`a`=start, `b`=count), the
    /// sketch-primitive node kinds, and `CurveDistanceAlong`'s
    /// primitive list (via `Intrinsic.primitive_start`/`primitive_count`).
    /// Replaces the legacy `primitives` side table.
    node_refs: [max_node_refs]i32 = undefined,
    node_ref_count: usize = 0,

    pub fn node(self: *MathIR, n: Node) !Expr {
        if (self.node_count >= max_nodes) return error.NodeCapacity;
        const id = self.node_count;
        self.nodes[id] = n;
        self.node_count += 1;
        return .{ .id = @intCast(id) };
    }

    pub fn var_(self: *MathIR, axis: Axis) !Expr {
        return self.node(.{ .kind = .var_, .op = @intFromEnum(axis) });
    }

    pub fn x(self: *MathIR) !Expr {
        return self.var_(.x);
    }
    pub fn y(self: *MathIR) !Expr {
        return self.var_(.y);
    }
    pub fn z(self: *MathIR) !Expr {
        return self.var_(.z);
    }
    pub fn slot(self: *MathIR, id: i32) !Expr {
        return self.node(.{ .kind = .slot, .op = id });
    }
    pub fn constant(self: *MathIR, value: f64) !Expr {
        return self.node(.{ .kind = .const_, .value = value });
    }
    pub fn unary(self: *MathIR, op: Unary, a: Expr) !Expr {
        return self.node(.{ .kind = .unary, .op = @intFromEnum(op), .a = a.id });
    }
    pub fn binary(self: *MathIR, op: Binary, a: Expr, b: Expr) !Expr {
        return self.node(.{ .kind = .binary, .op = @intFromEnum(op), .a = a.id, .b = b.id });
    }
    pub fn remapAxes(self: *MathIR, target: Expr, x_expr: Expr, y_expr: Expr, z_expr: Expr) !Expr {
        return self.node(.{ .kind = .remap_axes, .a = target.id, .b = x_expr.id, .c = y_expr.id, .d = z_expr.id });
    }
    pub fn affine3(self: *MathIR, affine: Affine3) !i32 {
        if (self.affine_count >= max_affines) return error.AffineCapacity;
        const id = self.affine_count;
        self.affines[id] = affine;
        self.affine_count += 1;
        return @intCast(id);
    }
    pub fn remapAffine(self: *MathIR, target: Expr, affine: i32) !Expr {
        return self.node(.{ .kind = .remap_affine, .a = target.id, .b = affine });
    }
    // The slot-indexed `point2` / `lineSegment` / `bezierQuadratic` /
    // `circle` builders and the legacy `sketchDistance` / `sketchPath`
    // intrinsics were retired in Phase 3. Use the `*N` builders below
    // for sketch primitives and `fold(.min, …)` for "min over a curve
    // list" — both emit pure subtree nodes.
    /// Variadic fold over `children`. Returns a `Fold` node whose `a`
    /// field is the start index of the children in `node_refs` and `b`
    /// is the count. `op` selects the fold operator (min/max/sum).
    pub fn fold(self: *MathIR, op: FoldOp, children: []const Expr) !Expr {
        if (self.node_ref_count + children.len > max_node_refs) return error.NodeRefCapacity;
        const start = self.node_ref_count;
        for (children) |child| {
            self.node_refs[self.node_ref_count] = child.id;
            self.node_ref_count += 1;
        }
        return self.node(.{ .kind = .fold, .op = @intFromEnum(op), .a = @intCast(start), .b = @intCast(children.len) });
    }

    /// Internal helper for primitive-as-node builders. Packs `children`
    /// into `node_refs` and pushes a node of `kind` with `op` carrying
    /// the plane (and any extra flag bits).
    fn primitiveNode(self: *MathIR, kind: NodeKind, plane: Plane, op_extra: i32, children: []const Expr) !Expr {
        if (self.node_ref_count + children.len > max_node_refs) return error.NodeRefCapacity;
        const start = self.node_ref_count;
        for (children) |child| {
            self.node_refs[self.node_ref_count] = child.id;
            self.node_ref_count += 1;
        }
        return self.node(.{
            .kind = kind,
            .op = @as(i32, @intFromEnum(plane)) * 2 + op_extra,
            .a = @intCast(start),
            .b = @intCast(children.len),
        });
    }

    pub fn lineSegmentN(self: *MathIR, plane: Plane, p0x: Expr, p0y: Expr, p1x: Expr, p1y: Expr) !Expr {
        return self.primitiveNode(.line_segment, plane, 0, &.{ p0x, p0y, p1x, p1y });
    }

    pub fn circleN(self: *MathIR, plane: Plane, cx: Expr, cy: Expr, r: Expr) !Expr {
        return self.primitiveNode(.circle, plane, 0, &.{ cx, cy, r });
    }

    pub fn bezierQuadraticN(self: *MathIR, plane: Plane, p0x: Expr, p0y: Expr, p1x: Expr, p1y: Expr, p2x: Expr, p2y: Expr) !Expr {
        return self.primitiveNode(.bezier_quadratic, plane, 0, &.{ p0x, p0y, p1x, p1y, p2x, p2y });
    }

    pub fn bezierCubicN(self: *MathIR, plane: Plane, p0x: Expr, p0y: Expr, p1x: Expr, p1y: Expr, p2x: Expr, p2y: Expr, p3x: Expr, p3y: Expr) !Expr {
        return self.primitiveNode(.bezier_cubic, plane, 0, &.{ p0x, p0y, p1x, p1y, p2x, p2y, p3x, p3y });
    }

    pub fn arcCenterN(self: *MathIR, plane: Plane, sx: Expr, sy: Expr, ex: Expr, ey: Expr, cx: Expr, cy: Expr, clockwise: bool) !Expr {
        const extra: i32 = if (clockwise) 1 else 0;
        return self.primitiveNode(.arc_center, plane, extra, &.{ sx, sy, ex, ey, cx, cy });
    }

    pub fn curveDistanceAlong(self: *MathIR, plane: Plane, primitives: []const Expr, ax: Expr, ay: Expr, az: Expr, flip: bool) !Expr {
        if (self.intrinsic_count >= max_intrinsics) return error.IntrinsicCapacity;
        if (self.node_ref_count + primitives.len > max_node_refs) return error.NodeRefCapacity;
        const ref_start = self.node_ref_count;
        for (primitives) |p| {
            self.node_refs[self.node_ref_count] = p.id;
            self.node_ref_count += 1;
        }
        const id = self.intrinsic_count;
        self.intrinsics[id] = .{
            .kind = .curve_distance_along,
            .plane = plane,
            .primitive_start = @intCast(ref_start),
            .primitive_count = @intCast(primitives.len),
            .flip = flip,
            .ax = ax.id,
            .ay = ay.id,
            .az = az.id,
        };
        self.intrinsic_count += 1;
        return self.node(.{ .kind = .intrinsic, .a = @intCast(id) });
    }
};
